### PR TITLE
13 task graphql

### LINKED ISSUE #13: GraphQL API for Tasks
Create a GraphQL api for handeling Tasks.

Tasks:
Get one task
Get all tasks? (will there be public tasks or are all private, meaning they can only be seen by the user who created it?)
Get tasks based on other criteria like, due date, categories, users, completed/uncompleted, ...?
Create a new task
Update/Change an existing task
Delete an existing task
Get all tasks (will there be public and private tasks?):

### LINKED ISSUE #16: Additional dependencies need for Flyway to work correctly
Adding following dependencies to pom.xml will fix it:

        <dependency>
            <groupId>org.flywaydb</groupId>
            <artifactId>flyway-mysql</artifactId>
            <version>9.21.0</version>
        </dependency>
        <dependency>
            <groupId>com.mysql</groupId>
            <artifactId>mysql-connector-j</artifactId>
            <scope>runtime</scope>
        </dependency>

And add this line to application-dev.properties:

` spring.flyway.baseline-on-migrate=true `

### PR DESCRIPTION
**Pull request by jakobpalmborg**
# GraphQL for Tasks

The following is implemented:

- Query
	- taskById
	- allTasks
	- tasksByCategoryId
	- tasksByTagId
- Mutation
	- addTask
	- deleteTask
	- updateTask

Test the API here:
http://localhost:8080/graphiql

- Dependencies for GraphQl
- Missing dependencies for Flyway
- Edit application-dev.properties
- Add Repesitories for Category, Tag and Task
- Add TaskController, TaskCreate, TaskService, TaskUpdate
- Add Schema for GraphQL
- Add data.sql
- Change Instant to LocalDateTime in Entities
- Add exceptions: CategoryNotFoundException, ExceptionResolver, ResourceNotFoundException, TagNotFoundException, TaskNotFoundException

This 
closes #13
closes #16

### CODE REVIEW

**Review by liber09 — APPROVED** [review]
Looks good as far as I can see

**Comment by a-jaxell** [comment]
Great job! Very clean to read and understand :)

I was just wondering if it would be preferable to somehow separate the Dtos in some way? Like adding Dto to their classname or store them in a folder named dto. 

In my opinion it would show their intended use more clearly when browsing the codebase.

**Comment by kappsegla** [comment]
What about tests? https://docs.spring.io/spring-graphql/reference/testing.html
Would feel much better if we have some automatic tests for the endpoints.

**Comment by solidtop** [comment]
> What about tests? https://docs.spring.io/spring-graphql/reference/testing.html Would feel much better if we have some automatic tests for the endpoints.



> Great job! Very clean to read and understand :)
> 
> I was just wondering if it would be preferable to somehow separate the Dtos in some way? Like adding Dto to their classname or store them in a folder named dto.
> 
> In my opinion it would show their intended use more clearly when browsing the codebase.

Should we have all dtos

> What about tests? https://docs.spring.io/spring-graphql/reference/testing.html Would feel much better if we have some automatic tests for the endpoints.

I agree! Maybe we should do the tests in a seperate PR/branch?

**Comment by solidtop** [comment]
> What about tests? https://docs.spring.io/spring-graphql/reference/testing.html Would feel much better if we have some automatic tests for the endpoints.

I agree! Maybe we should do the tests in a seperate PR/branch?

File: src/main/java/org/fungover/mmotodo/task/TaskCreateDto.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package org.fungover.mmotodo.task;
+
+import jakarta.validation.constraints.NotEmpty;
+import jakarta.validation.constraints.NotNull;
+
+public record TaskCreateDto(
```
Comment (a-jaxell) [inline src/main/java/org/fungover/mmotodo/task/TaskCreateDto.java:6]: 🥳

File: src/main/java/org/fungover/mmotodo/task/TaskUpdateDto.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package org.fungover.mmotodo.task;
+
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.PositiveOrZero;
+import org.springframework.boot.context.properties.bind.DefaultValue;
+
+public record TaskUpdateDto(
```
Comment (a-jaxell) [inline src/main/java/org/fungover/mmotodo/task/TaskUpdateDto.java:7]: 🥳

**Review by a-jaxell — APPROVED** [review]
Nice! Good job! 👏 

I had issues submitting a review for this unless I was assigned as a reviewer. After I assigned myself I could post a review. 

I dont know if this is a setting thats turned on because I havent experienced it before.

**Comment by kappsegla** [comment]
I have changed the implementation of TaskControllerTest. After adding a bean for LocalDateTimeScalar the test now runs as expected with:

@GraphQlTest
@Import(value = {GraphQlConfig.class})
class TaskControllerTest {

**Review by a-jaxell — APPROVED** [review]

**Review by kappsegla — APPROVED** [review]

**Pull request merged**